include/ugfx.h

```diff
@@ -0,0 +1,601 @@
+#define UGFX_ATOMIC_PRIM              1ULL << 55
+
+#define UGFX_MTX_STACK_MODEL     0x00ULL
+#define UGFX_MTX_STACK_VIEW_PROJ 0x01ULL
+#define UGFX_MTX_FORCE           0x02ULL
+#define UGFX_MTX_POP             0x04ULL
+#define UGFX_MTX_LOAD            0x00ULL
+#define UGFX_MTX_MUL             0x08ULL
+#define UGFX_MTX_NOPUSH          0x00ULL
+#define UGFX_MTX_PUSH            0x10ULL
+
+#define UGFX_CULL_NONE  0x0ULL
+#define UGFX_CULL_BACK  0x1ULL
+#define UGFX_CULL_FRONT 0x2ULL
+#define UGFX_CULL_BOTH  0x3ULL
+
+#define UGFX_GEOMETRY_FILL     0x00ULL
+#define UGFX_GEOMETRY_ZBUFFER  0x01ULL
+#define UGFX_GEOMETRY_TEXTURE  0x02ULL
+#define UGFX_GEOMETRY_SHADE    0x04ULL
+#define UGFX_GEOMETRY_SMOOTH   0x08ULL
+#define UGFX_GEOMETRY_LIGHTING 0x10ULL
+#define UGFX_GEOMETRY_FULL     UGFX_GEOMETRY_ZBUFFER | UGFX_GEOMETRY_TEXTURE | UGFX_GEOMETRY_SHADE | UGFX_GEOMETRY_SMOOTH | UGFX_GEOMETRY_LIGHTING
+
+#define UGFX_LINK_PUSH 0x0ULL
+#define UGFX_LINK_LOAD 0x1ULL
+
+#define UGFX_OP_NOOP                   0x00ULL
+#define UGFX_OP_TEXTURE_RECTANGLE      0x24ULL
+#define UGFX_OP_TEXTURE_RECTANGLE_FLIP 0x25ULL
+#define UGFX_OP_SYNC_PIPE              0x27ULL
+#define UGFX_OP_SYNC_TILE              0x28ULL
+#define UGFX_OP_SYNC_FULL              0x29ULL
+#define UGFX_OP_SET_KEY_GB             0x2AULL
+#define UGFX_OP_SET_KEY_R              0x2BULL
+#define UGFX_OP_SET_CONVERT            0x2CULL
+#define UGFX_OP_SET_SCISSOR            0x2DULL
+#define UGFX_OP_SET_PRIM_DEPTH         0x2EULL
+#define UGFX_OP_SET_OTHER_MODES        0x2FULL
+#define UGFX_OP_LOAD_TLUT              0x30ULL
+#define UGFX_OP_SYNC_LOAD              0x31ULL
+#define UGFX_OP_SET_TILE_SIZE          0x32ULL
+#define UGFX_OP_LOAD_BLOCK             0x33ULL
+#define UGFX_OP_LOAD_TILE              0x34ULL
+#define UGFX_OP_SET_TILE               0x35ULL
+#define UGFX_OP_FILL_RECTANGLE         0x36ULL
+#define UGFX_OP_SET_FILL_COLOR         0x37ULL
+#define UGFX_OP_SET_FOG_COLOR          0x38ULL
+#define UGFX_OP_SET_BLEND_COLOR        0x39ULL
+#define UGFX_OP_SET_PRIM_COLOR         0x3AULL
+#define UGFX_OP_SET_ENV_COLOR          0x3BULL
+#define UGFX_OP_SET_COMBINE_MODE       0x3CULL
+#define UGFX_OP_SET_TEXTURE_IMAGE      0x3DULL
+#define UGFX_OP_SET_Z_IMAGE            0x3EULL
+#define UGFX_OP_SET_COLOR_IMAGE        0x3FULL
+#define UGFX_OP_FINALIZE               0x80ULL
+#define UGFX_OP_LOAD_VERTICES          0x81ULL
+#define UGFX_OP_LOAD_MATRIX            0x82ULL
+#define UGFX_OP_LOAD_VIEWPORT          0x83ULL
+#define UGFX_OP_LOAD_LIGHT             0x84ULL
+#define UGFX_OP_LINK_COMMANDS          0x85ULL
+#define UGFX_OP_SET_CULL_MODE          0x86ULL
+#define UGFX_OP_SET_GEOMETRY_MODE      0x87ULL
+#define UGFX_OP_SET_PERSP_NORM         0x88ULL
+#define UGFX_OP_SET_TEXTURE_SETTINGS   0x89ULL
+#define UGFX_OP_SET_NUM_LIGHTS         0x8AULL
+#define UGFX_OP_SET_CLIP_RATIO         0x8BULL
+#define UGFX_OP_SET_ADDRESS_SLOT       0x8CULL
+#define UGFX_OP_DRAW_TRIANGLE          0x8DULL
```

Comment:
I'm a big fan of this. In libultra, the microcode needs to be re-uploaded for every single task, which seems wasteful. This idea of a modular microcode system sounds a lot better to me. It also sounds like a lot of design effort needs to go into this, so a big question is how we should approach this. I saw that you opened another PR with a new audio system (which looks great by the way) that also includes microcode. Would you like to merge that first, or shall we try and make the two systems compatible right away? In any case, I'd gladly participate in the effort.

Reply:
I think it would be better to review and land that PR and this one independently, and then at some point work towards the unified command list. It could happen even after we merge both -- at the end of the day, if you think of it, it's all behind the public API (once you hide the explicit display list management).

I think your PR is a better staging area for this because the graphics microcode already needs to parse a command list (the audio microcode that I submitted would be a single command, I think -- but I have more to submit that would require different commands).